### PR TITLE
Allow declaring and passing resources to conditions

### LINKED ISSUE #1021: Conditionals: Allow access to pipeline resources
# Expected Behavior
Condition containers should have access to the resources themselves e.g a `fileExists` condition to check if a file exists in a git repo. 

# Additional Info
The resources that the condition needs have to be explicitly specified so that we do not needlessly initialize all resources for each condition check. Based on design in #27 

ref: https://github.com/tektoncd/pipeline/issues/1137

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

Allow condition to declare pipeline resources in their definitions.
The resource can be accessed by the condition container during a check.
Only input resources are supported. The declared resources can be specified
in a PipelineTaskCondition and variable substitution for the resource is
supported just like with a `TaskResource`.

Fixes #1021 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

# Release Notes

```
Adds resource support for conditionals. Conditions have to declare 
the resources they can use and these can be passed in via the
Pipeline and PipelineRun spec similar to TaskResources. 
```

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/condition_types.go [hunk 1]
```diff
@@ -73,6 +73,26 @@ type ConditionSpec struct {
...
+type ConditionResource struct {
+	// Name declares the name by which a resource is referenced in the Condition's
+	// definition.
+	Name string `json:"name"`
+	// Type is the type of this resource;
+	Type PipelineResourceType `json:"type"`
+	// TargetPath is the path in workspace directory where the resource
+	// will be copied.
+	// +optional
+	TargetPath string `json:"targetPath"`
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/condition_types.go]: when i first saw this i wondered, why not use the existing type?

https://github.com/tektoncd/pipeline/blob/bb855711f077b43c7948f37f5406bff69b4b29de/pkg/apis/pipeline/v1alpha1/task_types.go#L109-L122

I'm guessing it's b/c of the `OutputImageDir` field on the condition resource, is that correct? if that's the only reason then im thinking maybe we could do something like:

- Define one type (maybe called something like `ResourceInput`? or something better XD)
- Use this new type directly for conditions
- Make it so that `TaskResource` embeds the type and adds `OutputImageDir`
- Maybe open an issue to explore changing `OutputImageDir` (I'm not a big fan of the fact that the generic struct `TypeResource` has an image specific field :S)

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -626,6 +626,11 @@ func (c *Reconciler) makeConditionCheckContainer(rprt *resources.ResolvedPipelin
 	labels := getTaskrunLabels(pr, rprt.PipelineTask.Name)
 	labels[pipeline.GroupName+pipeline.ConditionCheckKey] = rcc.ConditionCheckName
 
+	taskSpec, err := rcc.ConditionToTaskSpec()
+	if err != nil {
+		return nil, err
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: its often helpful to augment the error with some context before returning it, e.g. ` xerrors.Errorf("Failed to get TaskSpec from condition: %w", err)`

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 3]
```diff
@@ -95,24 +99,65 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
...
+	err := convertResourceTemplateStrings(&t.Steps[0], rcc.ResolvedResources, rcc.Condition.Spec.Resources)
+
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to replace resource template strings %w", err)
+	}
+
+	return t, nil
+}
+
+func toTaskResouces(resources []v1alpha1.ConditionResource) []v1alpha1.TaskResource {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: might not need this if we can use the same type (ish?? see my comment above) for both this and TaskResources :D

Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: if we _do_ still need this function - maybe it would make sense to have this as a function on the `ConditionResource` itself, in the `_types.go` file?

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 4]
```diff
@@ -95,24 +99,65 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
...
 	for _, p := range params {
 		replacements[fmt.Sprintf("params.%s", p.Name)] = fmt.Sprintf("${inputs.params.%s}", p.Name)
-		templating.ApplyContainerReplacements(container, replacements, map[string][]string{})
 	}
+
+	templating.ApplyContainerReplacements(container, replacements, map[string][]string{})
+}
+
+// Prepends inputs. to all resource template strings so that they can be replaced by
+// taskrun variable substitution e.g. ${resources.name.key} to ${inputs.resources.name.key}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: hm this seems like a bit of a hack 😇 im thinking in the long run we should refactor this so we can call the `templating` (now called `substitution` 😅 ) functions without needing this hack first (np if we want to do this in a separate PR or even open an issue about it)

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 5]
```diff
@@ -95,24 +99,65 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
...
+func convertResourceTemplateStrings(container *corev1.Container,
+	resolvedResources map[string]*v1alpha1.PipelineResource,
+	conditionResources []v1alpha1.ConditionResource) error {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: this indentation seems kinda non - standard, makes it a bit hard to read - i think it's okay to just have a really long line for the function signature imo

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 6]
```diff
@@ -95,24 +99,65 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
...
+	replacements := make(map[string]string)
+	for _, cr := range conditionResources {
+		if rSpec, ok := resolvedResources[cr.Name]; ok {
+			r, err := v1alpha1.ResourceFromType(rSpec)
+			if err != nil {
+				return err
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: could add some context to this error, as mentioned above

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go [hunk 7]
```diff
@@ -467,12 +468,35 @@ func resolveConditionChecks(pt *v1alpha1.PipelineTask,
...
+				// Next, fetch the actual resource definition
+				gotResource, err := getResource(resourceRef.Name)
+				if err != nil {
+					return nil, xerrors.Errorf("could not retrieve resource %s referenced in condition %s for pipeline task %s", r.Name, cName, pt.Name)
+				}
+
+				// Finally add it to the resolved resources map
+				rr[r.Name] = gotResource
+			}
+			rcc.ResolvedResources = rr
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: this kinda feels like it could be a separate function, what do you think? 

`resolveConditionChecks` could even become `ResolveConditionChecks` and be called by the reconciler after calling `ResolvePipelineRun` - that would make it so `ResolvePipelineRun` and its unit tests could be simpler, and you'd no longer need to distinguish between different kinds of errors

File: pkg/apis/pipeline/v1alpha1/condition_types.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/apis/pipeline/v1alpha1/condition_types.go]: > Maybe open an issue to explore changing OutputImageDir (I'm not a big fan of the fact that the generic struct TypeResource has an image specific field :S)

We can probably leave that for now - between [the resource extensibility design](https://docs.google.com/document/d/1rcMG1cIFhhixMSmrT734MBxvH3Ghre9-4S2wbzodPiU) and [a proposal im making for treating output images differently](https://docs.google.com/document/d/1fF2vWMs12d3FwkqkNuzS7FzQFQSFz1ZKUe6CHqkIg0c/edit#heading=h.waoz4clirr5h) i think we're gonna get it fixed :D

but in the meantime embedding might be an okay solution!

**Comment by bobcatfish** [comment]
```
I0816 14:00:38.929] ERROR: test demo-condtional-pr-list-files-1-gknrc-file-exists-68bvx=SucceededFalse but should be succeededtrue
I0816 14:00:38.930] ERROR: one or more YAML tests failed
```

ooo that might be legit

**Comment by afrittoli** [comment]
Thanks for this! I will review it now. A quick note, there are three commits in this PR, do you plan to squash or if not could you amend the commit messages of the last two commits?

**Comment by dibyom** [comment]
@afrittoli Thanks...Yes, will squash and also going to make a couple of changes later today based on @bobcatfish 's review comment.

File: docs/conditions.md [hunk 8]
```diff
@@ -57,10 +58,31 @@ only start with alpha characters and `_`. For example, `fooIs-Bar_` is a valid
...
 multiple protected branch names. When the actual parameter value is supplied, its parsed type 
 is validated against the type field.
 
+### Resources
+
+Conditions can declare input [`PipelineResources`](resources.md)  via the `resources` field to 
+provide the Condition container step with data or context that is needed to perform the check.
+
+Input resources, like source code (git), are dumped at path
+`/workspace/resource_name` within a mounted
```
[the remark this answers is not in the document]
Comment (dibyom) [reply docs/conditions.md]: Yeah, the thing is since things like `targetPath` are part of `TaskResource`, they are in the tasks doc. Now with conditions, there is some duplication. I'll add a section in the resource doc on how to use resources in conditions and tasks. That section will link to both the Task and Condition resource docs and contain the common bits e.g. targetPaths, the path variable substitution.

I'm thinking of doing this in a follow-up PR.

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go (hunk 4 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: done!

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go (hunk 7 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: ok, I split the resource resolution into a separate function. There is definitely a lot of scope to refactor the pipelineState resolution -- besides resolving the conditionChecks separately, another thing that might help (particularly with #1023) would be changing the `PipelineState` from just being a slice of the ResolvedTasks to  having a reference to the `Dag` nodes.

Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: > There is definitely a lot of scope to refactor the pipelineState resolution

Yeah probably too much to do all at ounce - but thanks for the incremental improvement!

File: pkg/apis/pipeline/v1alpha1/pipeline_types.go [hunk 9]
```diff
@@ -135,6 +138,15 @@ type PipelineDeclaredResource struct {
 	Type PipelineResourceType `json:"type"`
 }
 
+// PipelineConditionResource allows a Pipeline to declare how its DeclaredPipelineResources
+// should be provided to a Condition as its inputs.
+type PipelineConditionResource struct {
+	// Name is the name of the PipelineResource as declared by the Condition.
+	Name string `json:"name"`
+	// Resource is the name of the DeclaredPipelineResource to use.
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/pipeline_types.go]: hm this is kinda odd but also really minor - these comments refer to `DeclaredPipelineResource` but I don't think that `DeclaredPipelineResource` is a type that's actually defined anywhere?

File: pkg/apis/pipeline/v1alpha1/resource_paths_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	}}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			if actual := v1alpha1.OutputResourcePath(tc.resource); actual != tc.expected {
+				t.Errorf("Unexpected output resource path: %s", actual)
+			}
+		})
+	}
+}
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/resource_paths_test.go]: haha yeessssss i love the super focused test coverage XD

File: pkg/apis/pipeline/v1alpha1/resource_paths_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+			Name:       "foo",
+			TargetPath: "bar",
+		},
+		expected: "/workspace/bar",
+	}}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			if actual := v1alpha1.OutputResourcePath(tc.resource); actual != tc.expected {
+				t.Errorf("Unexpected output resource path: %s", actual)
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/resource_paths_test.go]: can be helpful to included the expected path in the error message as well

File: pkg/apis/pipeline/v1alpha1/task_types.go [hunk 12]
```diff
@@ -117,17 +117,7 @@ type Inputs struct {
...
-	// Task's steps.
-	Name string `json:"name"`
-	// Type is the type of this resource;
-	Type PipelineResourceType `json:"type"`
-	// TargetPath is the path in workspace directory where the task resource
-	// will be copied.
-	// +optional
-	TargetPath string `json:"targetPath,omitempty"`
-	// Path to the index.json file for output container images.
+	ResourceDeclaration
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/task_types.go]: :tada:  :innocent:

File: pkg/apis/pipeline/v1alpha1/pipeline_types.go (hunk 9 above)
Comment (bobcatfish) [reply pkg/apis/pipeline/v1alpha1/pipeline_types.go]: (Maybe this is `ResourceDeclaration`)?

**Comment by dibyom** [comment]
> 

pkg/reconciler/pipelinerun/pipelinerun.go | 94.4% | 82.2% | -12.2
-- | -- | -- | --


:thinking:

**Comment by dibyom** [comment]
> I just saw the -12.2% coverage thing so maybe I'll wait for that XD

Not sure this is accurate. The coverage for that file is around 82.2% according to another recently updated PR: https://github.com/tektoncd/pipeline/pull/1204#issuecomment-522044106

Trying to run this locally fails with a data race :( 
(I think this is the same issue as #1124)

**Comment by bobcatfish** [comment]
Something might be up with that coverage bot anyway #1248
